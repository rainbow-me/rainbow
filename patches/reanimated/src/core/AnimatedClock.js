import InternalAnimatedValue from './InternalAnimatedValue';
import AnimatedNode from './AnimatedNode';
import { val } from '../val';

class AnimatedMainClock extends InternalAnimatedValue {
  _frameCallback;

  constructor() {
    super({ type: 'MAIN_CLOCK' });
  }

  __onEvaluate() {
    return +new Date();
  }

  _runFrame = () => {
    this._updateValue(0);
    if (this.__children.length > 0) {
      this._frameCallback = requestAnimationFrame(this._runFrame);
    }
  };

  __attach() {
    super.__attach();
    if (!this._frameCallback) {
      this._frameCallback = requestAnimationFrame(this._runFrame);
    }
  }

  __detach() {
    if (this._frameCallback) {
      cancelAnimationFrame(this._frameCallback);
      this._frameCallback = null;
    }
    super.__detach();
  }
}

const mainClock = new AnimatedMainClock();

export default class AnimatedClock extends AnimatedNode {
  _started;
  _attached;

  constructor() {
    super({ type: 'clock' });
  }

  toString() {
    return `AnimatedClock, id: ${this.__nodeID}`;
  }

  __onEvaluate() {
    return val(mainClock);
  }

  __attach() {
    super.__attach();
    if (this._started && !this._attached) {
      mainClock.__addChild(this);
    }
    this._attached = true;
  }

  __detach() {
    if (this._started && this._attached) {
      mainClock.__removeChild(this);
    }
    this._attached = false;
    super.__detach();
  }

  start() {
    if (!this._started && this._attached) {
      mainClock.__addChild(this);
    }
    this._started = true;
  }

  stop() {
    if (this._started && this._attached) {
      mainClock.__removeChild(this);
    }
    this._started = false;
  }

  isStarted() {
    return this._started;
  }
}

