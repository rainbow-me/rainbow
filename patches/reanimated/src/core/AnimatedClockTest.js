import AnimatedNode from './AnimatedNode';

class AnimatedClockTest extends AnimatedNode {
  _clockNode;

  constructor(clockNode) {
    super({ type: 'clockTest', clock: clockNode });
    this._clockNode = clockNode;
  }

  toString() {
    return `AnimatedClockTest, id: ${this.__nodeID}`;
  }

  __onEvaluate() {
    return this._clockNode.isStarted() ? 1 : 0;
  }
}

export function createAnimatedClockTest(clock) {
  return new AnimatedClockTest(clock);
}
