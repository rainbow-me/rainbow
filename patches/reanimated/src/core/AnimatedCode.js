import useCode from '../derived/useCode';
import AnimatedNode from './AnimatedNode';

function assertNodesNotNull(code, children, exec) {
  if (!code) {
    const error = !children
      ? `Got "${typeof children}" type passed to children`
      : `Got "${typeof exec}" type passed to exec`;

    throw new Error(
      `<Animated.Code /> expects the 'exec' prop or children to be an animated node or a function returning an animated node. ${error}`
    );
  }
}

function Code({ exec, children, dependencies }) {
  const nodes = children || exec;

  let code = null;
  if (nodes instanceof AnimatedNode) {
    code = () => nodes;
  } else if (typeof nodes === 'function') {
    code = nodes;
  }

  assertNodesNotNull(code, children, exec);

  useCode(code, dependencies);
  return null;
}

export default Code;
