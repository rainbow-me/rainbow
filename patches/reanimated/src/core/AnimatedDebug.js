import invariant from 'fbjs/lib/invariant';
import { val } from '../val';
import { adapt, createAnimatedBlock as block } from './AnimatedBlock';
import { createAnimatedCall as call } from './AnimatedCall';
import AnimatedNode from './AnimatedNode';

class AnimatedDebug extends AnimatedNode {
  _message;
  _value;

  constructor(message, value) {
    invariant(
      typeof message === 'string',
      `Reanimated: Animated.debug node first argument should be of type string but got ${message}`
    );
    invariant(
      value instanceof AnimatedNode,
      `Reanimated: Animated.debug node second argument should be of type AnimatedNode but got ${value}`
    );
    super({ type: 'debug', message, value }, [value]);
    this._message = message;
    this._value = value;
  }

  toString() {
    return `AnimatedDebug, id: ${this.__nodeID}`;
  }

  __onEvaluate() {
    const value = val(this._value);
    console.log(this._message, value);
    return value;
  }
}

export function createAnimatedDebug(message, value) {
  if (__DEV__) {
    const runningInRemoteDebugger = typeof atob !== 'undefined';
    // hack to detect if app is running in remote debugger
    // https://stackoverflow.com/questions/39022216

    const runningInExpoShell =
      global.Expo && global.Expo.Constants.appOwnership !== 'standalone';

    if (runningInRemoteDebugger || runningInExpoShell) {
      // When running in expo or remote debugger we use JS console.log to output variables
      // otherwise we output to the native console using native debug node
      return block([
        call([value], ([a]) => console.log(`${message} ${a}`)),
        value,
      ]);
    } else {
      return new AnimatedDebug(message, adapt(value));
    }
  }
  // Debugging is disabled in PROD
  return value;
}
