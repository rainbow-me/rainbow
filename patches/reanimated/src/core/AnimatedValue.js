import { createAnimatedSet as set } from './AnimatedSet';
import interpolate from '../derived/interpolate';
import InternalAnimatedValue from './InternalAnimatedValue';
import { Platform } from 'react-native';
import { evaluateOnce } from '../derived/evaluateOnce';
import ReanimatedModule from '../ReanimatedModule';
import { val } from '../val';

// Animated value wrapped with extra methods for omit cycle of dependencies
export default class AnimatedValue extends InternalAnimatedValue {
  setValue(value) {
    this.__detachAnimation(this._animation);
    if (Platform.OS === 'web' || Platform.OS === 'windows' || Platform.OS === 'macos') {
      this._updateValue(val(value));
    } else {
      if (ReanimatedModule.setValue && typeof value === 'number') {
        // FIXME Remove it after some time
        // For OTA-safety
        // FIXME handle setting value with a node
        ReanimatedModule.setValue(this.__nodeID, value);
      } else {
        evaluateOnce(set(this, value), this);
      }
    }
  }

  toString() {
    return `AnimatedValue, id: ${this.__nodeID}`;
  }

  interpolate(config) {
    return interpolate(this, config);
  }
}
