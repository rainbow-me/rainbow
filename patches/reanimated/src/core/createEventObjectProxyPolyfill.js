// JSC on Android and iOS 8 & 9 does not support proxies.
// The below workaround provides a proxy-like functionality for event data by
// creating an object which contains predefined fields for the most commonly
// used event attributes. If your app uses even attribute which is not listed
// here please submit a PR to add that and we will merge it right away
export default function createEventObjectProxyPolyfill() {
  const nodesMap = {
    // Gesture handlers-related event
    translationX: {},
    translationY: {},
    state: {},
    oldState: {},
    absoluteX: {},
    absoluteY: {},
    x: {},
    y: {},
    velocityX: {},
    velocityY: {},
    scale: {},
    focalX: {},
    focalY: {},
    rotation: {},
    anchorX: {},
    anchorY: {},
    velocity: {},
    numberOfPointers: {},
    // onLayour-related event
    layout: { x: {}, y: {}, width: {}, height: {} },
    // ScrollView event
    contentOffset: { y: {}, x: {} },
    layoutMeasurement: { width: {}, height: {} },
    contentSize: { width: {}, height: {} },
    zoomScale: {},
    contentInset: { right: {}, top: {}, left: {}, bottom: {} },
  };
  const traverse = obj => {
    for (const key in obj) {
      traverse(obj[key]);
      Object.assign(obj[key], { __isProxy: true });
    }
  };
  traverse(nodesMap);
  return nodesMap;
}
