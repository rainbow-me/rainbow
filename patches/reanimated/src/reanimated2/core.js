/* global _WORKLET */

import NativeReanimated from './NativeReanimated';

global.__reanimatedWorkletInit = function(worklet) {
  worklet.__worklet = true;
};

function pushFrame(frame) {
  NativeReanimated.pushFrame(frame);
}

export function requestFrame(frame) {
  'worklet';

  if (NativeReanimated.native) {
    requestAnimationFrame(frame);
  } else {
    pushFrame(frame);
  }
}

global._WORKLET = false;
global._log = function(s) {
  console.log(s);
};

export function runOnUI(worklet) {
  return makeShareable(worklet);
}

export function makeShareable(value) {
  return NativeReanimated.makeShareable(value);
}

export function getViewProp(viewTag, propName) {
  return new Promise((resolve, reject) => {
    return NativeReanimated.getViewProp(viewTag, propName, (result) => {
      if (result.substr(0, 6) === 'error:') {
        reject(result);
      } else {
        resolve(result);
      }
    });
  });
}

function workletValueSetter(value) {
  'worklet';
  const previousAnimation = this._animation;
  if (previousAnimation) {
    previousAnimation.cancelled = true;
    this._animation = null;
  }
  if (
    typeof value === 'function' ||
    (value !== null && typeof value === 'object' && value.onFrame)
  ) {
    // animated set
    const animation = typeof value === 'function' ? value() : value;
    let initializeAnimation = (timestamp) => {
      animation.onStart(animation, this.value, timestamp, previousAnimation);
    };
    const step = (timestamp) => {
      if (animation.cancelled) {
        animation.callback && animation.callback(false /* finished */);
        return;
      }
      if (initializeAnimation) {
        initializeAnimation(timestamp);
        initializeAnimation = null; // prevent closure from keeping ref to previous animation
      }
      const finished = animation.onFrame(animation, timestamp);
      animation.timestamp = timestamp;
      this._value = animation.current;
      if (finished) {
        animation.callback && animation.callback(true /* finished */);
      } else {
        requestAnimationFrame(step);
      }
    };

    this._animation = animation;

    requestAnimationFrame(step);
  } else {
    this._value = value;
  }
}

// We cannot use pushFrame
// so we use own implementation for js
function workletValueSetterJS(value) {
  const previousAnimation = this._animation;
  if (previousAnimation) {
    previousAnimation.cancelled = true;
    this._animation = null;
  }
  if (
    typeof value === 'function' ||
    (value !== null && typeof value === 'object' && value.animation)
  ) {
    // animated set
    const animation = typeof value === 'function' ? value() : value;
    let initializeAnimation = (timestamp) => {
      animation.onStart(animation, this.value, timestamp, previousAnimation);
    };
    const step = (timestamp) => {
      if (animation.cancelled) {
        animation.callback && animation.callback(false /* finished */);
        return;
      }
      if (initializeAnimation) {
        initializeAnimation(timestamp);
        initializeAnimation = null; // prevent closure from keeping ref to previous animation
      }
      const finished = animation.onFrame(animation, timestamp);
      animation.timestamp = timestamp;
      this._setValue(animation.current);
      if (finished) {
        animation.callback && animation.callback(true /* finished */);
      } else {
        requestFrame(step);
      }
    };

    this._animation = animation;

    requestFrame(step);
  } else {
    this._setValue(value);
  }
}

NativeReanimated.installCoreFunctions(
  NativeReanimated.native ? workletValueSetter : workletValueSetterJS
);

export function makeMutable(value) {
  return NativeReanimated.makeMutable(value);
}

export function makeRemote(object = {}) {
  return NativeReanimated.makeRemote(object);
}

export function startMapper(mapper, inputs = [], outputs = []) {
  return NativeReanimated.startMapper(mapper, inputs, outputs);
}

export function stopMapper(mapperId) {
  NativeReanimated.stopMapper(mapperId);
}

export const runOnJS = (fun) => {
  'worklet';
  if (!_WORKLET) {
    return fun;
  }
  if (!fun.__callAsync) {
    throw new Error(
      "Attempting to call runOnJS with an object that is not a host function. Using runOnJS is only possible with methods that are defined on the main React-Native Javascript thread and that aren't marked as worklets"
    );
  } else {
    return fun.__callAsync;
  }
};

const capturableConsole = console;
runOnUI(() => {
  'worklet';
  const console = {
    log: runOnJS(capturableConsole.log),
    warn: runOnJS(capturableConsole.warn),
    error: runOnJS(capturableConsole.error),
  };
  _globalSetter('console', console); // eslint-disable-line
})();
