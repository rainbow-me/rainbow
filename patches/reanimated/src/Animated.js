import { Image, ScrollView, Text, View } from 'react-native';
import EasingNode from './Easing';
import AnimatedClock from './core/AnimatedClock';
import AnimatedValue from './core/AnimatedValue';
import AnimatedNode from './core/AnimatedNode';
import AnimatedCode from './core/AnimatedCode';
import * as base from './base';
import * as derived from './derived';
import createAnimatedComponent from './createAnimatedComponent';
import decay from './animations/decay';
import timing from './animations/timing';
import spring from './animations/spring';
import Animation from './animations/Animation';
import {
  addWhitelistedNativeProps,
  addWhitelistedUIProps,
} from './ConfigHelper';
import backwardCompatibleAnimWrapper from './animations/backwardCompatibleAnimWrapper';
import {
  Transition,
  Transitioning,
  createTransitioningComponent,
} from './Transitioning';
import SpringUtils from './animations/SpringUtils';
import useValue from './useValue';
import * as reanimated2 from './reanimated2';

const decayWrapper = backwardCompatibleAnimWrapper(
  decay,
  Animation.decayDefaultState
);
const timingWrapper = backwardCompatibleAnimWrapper(
  timing,
  Animation.timingDefaultState
);
const springWrapper = backwardCompatibleAnimWrapper(
  spring,
  Animation.springDefaultState
);
const Animated = {
  // components
  View: createAnimatedComponent(View),
  Text: createAnimatedComponent(Text),
  Image: createAnimatedComponent(Image),
  ScrollView: createAnimatedComponent(ScrollView),
  Code: AnimatedCode,
  createAnimatedComponent,

  // classes
  Clock: AnimatedClock,
  Value: AnimatedValue,
  Node: AnimatedNode,

  // operations
  ...base,
  ...derived,

  // animations
  decay: decayWrapper,
  timing: timingWrapper,
  spring: springWrapper,
  SpringUtils,

  // configuration
  addWhitelistedNativeProps,
  addWhitelistedUIProps,

  // hooks
  useValue,

  // reanimated2
  ...reanimated2,
};

export default Animated;

// operations
export * from './base';
export * from './derived';

export * from './reanimated2';

export {
  EasingNode as Easing,
  Transitioning,
  Transition,
  createTransitioningComponent,
  // classes
  AnimatedClock as Clock,
  AnimatedValue as Value,
  AnimatedNode as Node,
  // animations
  decayWrapper as decay,
  timingWrapper as timing,
  springWrapper as spring,
  SpringUtils,
  // hooks
  useValue,
};
