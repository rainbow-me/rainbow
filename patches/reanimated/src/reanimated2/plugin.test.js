import plugin from '../../plugin';
import { transform } from '@babel/core';

function runPlugin(input) {
  return transform(input, {
    filename: 'someFile.js',
    compact: false,
    plugins: [plugin],
  });
}

describe('babel plugin', () => {
  it('transforms ', () => {
    const input = `
    import Animated, { useAnimatedStyle, useSharedValue } from 'react-native-reanimated'
    
    function Box() {
        const offset = useSharedValue(0);
      
        const animatedStyles = useAnimatedStyle(() => {
          return {
            transform: [{ translateX: offset.value * 255 }],
          };
        });
      
        return (
          <>
            <Animated.View style={[styles.box, animatedStyles]} />
            <Button onPress={() => (offset.value = Math.random())} title="Move" />
          </>
        );
      }
    `;

    const { code } = runPlugin(input);
    expect(code).toMatchSnapshot();
  });

  it('adds closure to worklets', () => {
    const input = `
      function someWorklet(greeting) {
        'worklet';
        console.log("Hey I'm running on the UI thread");
      }`;

    const { code } = runPlugin(input);
    expect(code).toContain('_closure');
  });

  it('supports default ES6 style imports', () => {
    const input = `
    import * as Reanimated from 'react-native-reanimated'

    function Box() {
        const offset = Reanimated.useSharedValue(0);

        const animatedStyles = Reanimated.useAnimatedStyle(() => {
          return {
            transform: [{ translateX: offset.value * 255 }],
          };
        });

        return (
          <>
            <Animated.View style={[styles.box, animatedStyles]} />
            <Button onPress={() => (offset.value = Math.random())} title="Move" />
          </>
        );
      }
    `;

    const { code } = runPlugin(input);
    expect(code).toContain('_closure');
  });
});
