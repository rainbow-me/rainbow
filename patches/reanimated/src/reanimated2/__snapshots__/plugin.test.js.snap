// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel plugin transforms  1`] = `
"var _interopRequireWildcard = require(\\"@babel/runtime/helpers/interopRequireWildcard\\");

var _reactNativeReanimated = _interopRequireWildcard(require(\\"react-native-reanimated\\"));

function Box() {
  var offset = (0, _reactNativeReanimated.useSharedValue)(0);
  var animatedStyles = (0, _reactNativeReanimated.useAnimatedStyle)(function () {
    const _f = function () {
      return {
        transform: [{
          translateX: offset.value * 255
        }]
      };
    };

    _f._closure = {
      offset
    };
    _f.asString = \\"function _f(){const{offset}=jsThis._closure;{return{transform:[{translateX:offset.value*255}]};}}\\";
    _f.__workletHash = 7114514849439;

    global.__reanimatedWorkletInit(_f);

    return _f;
  }());
  return React.createElement(React.Fragment, null, React.createElement(_reactNativeReanimated.default.View, {
    style: [styles.box, animatedStyles]
  }), React.createElement(Button, {
    onPress: function onPress() {
      return offset.value = Math.random();
    },
    title: \\"Move\\"
  }));
}"
`;
